### ISSUE TITLE
Control messages are scheduled every millisecond

### ISSUE THREAD

**Issue by mratsim** [status-im/nim-libp2p#386]
Discussion with @dryajov, currently the control messages seem to be sent every milliseconds. It might be because a backlog of them accumulated and the outdated one were not dropped:

![image](https://user-images.githubusercontent.com/22738317/94831275-d1265100-040c-11eb-9ab4-0739671622f2.png)

**Comment by dryajov** [comment]
Actually, control messages are sent to peers every second, but they are sent to every peer so it is possible to end up with bursts of control messages like the ones in the pic above. 

One way that the spec recommends dealing with this is piggybacking control messages on regular messages. 

Another way would be to have an intermediary FIFO queue, that would accumulate messages control messages and drop the ones that are past some threshold, for example 3 seconds... The queue is needed either way for piggybacking.

Overall, this is pretty unoptimized right now and needs a bit more thought put into it.

**Comment by sinkingsugar** [comment]
They are scheduled for n peers * n topics, that's what you are seeing.

**Comment by sinkingsugar** [comment]
@dryajov gossip 1.1 already set hard limits to avoid spam, and this is implemented.
I don't see a big issue, piggybacking is a good idea tho.

**Comment by arnetheduck** [comment]
well, one trivial thing to fix would be to ensure that the control message is sent per peer with all topics collected in it already instead of doing per peer per topic

**Comment by sinkingsugar** [comment]
We already do that. That specific trace comes from incoming RPCs, basically it's part of the re-broadcasting of messages, that's why the aggregation is not trivial @arnetheduck. We actually don't go thru topics (I was wrong).

https://github.com/status-im/nim-libp2p/blob/4a98a8af5a6b33e2c01ff34321c0127386037f79/libp2p/protocols/pubsub/gossipsub.nim#L1108-L1114

**Comment by sinkingsugar** [comment]
> One way that the spec recommends dealing with this is piggybacking control messages on regular messages.
> 
> Another way would be to have an intermediary FIFO queue, that would accumulate control messages and drop the ones that are past some threshold, for example 3 seconds... The queue is needed either way for piggybacking.
> 

This is probably the only fair solution.
Is it a major bottleneck tho? What is the problem here?

**Comment by dryajov** [comment]
> Is it a major bottleneck tho? What is the problem here?

We should measure it and see what's our message frequency and throughput overall. The overall amount of bytes sent is probably (about) the same, weather you write this messages at once or one by one, however encoding and decoding the messages has an added cost for both the local and remote nodes. The only real measurable difference would be extra varint headers sent per each extra frame and whatever additional overhead protobuf adds in terms of size.

Regarding the message bursts, I now realize that the message they are probably a red herring. Take a look at the heartbeat method https://github.com/status-im/nim-libp2p/blob/master/libp2p/protocols/pubsub/gossipsub10.nim#L274-L305, in particular at https://github.com/status-im/nim-libp2p/blob/master/libp2p/protocols/pubsub/gossipsub10.nim#L305, _we won't schedule another heartbeat unless the current one has ended_. 

The flow is:

1. get some random peers and some messages
2. write the messages to all the peers in step 1
3. sleep for `GossipSubHeartbeatInterval`
4. repeat from `1`

In other words, the interval is _at least_ `GossipSubHeartbeatInterval` second, but in practice, it's `GossipSubHeartbeatInterval` + `whatever work needs to be done prior to sleeping`.

That said, piggybacking is probably still a good improvement overall, but I wouldn't expect any sort of drastic perf or size gains.

**Comment by sinkingsugar** [comment]
The traces and this whole issue is not about `heartbeat` tho, but rather `rpcHandler`. Those are other peers rpcs broadcasts.

The heartbeat is indeed not exact real time but I don't think we have such requirement.

**Comment by sinkingsugar** [comment]
Indeed tho I'm surprised that the 1.1 spec did not touch this topic of re broadcasting. I will have a double check. Looks like a ddos vector after all.

**Comment by Menduist** [comment]
Closing in favor of #633
